contracts/sol6/fprConversionRates/QtyStepConversionRates.sol

```diff
@@ -0,0 +1,403 @@
+            imbalanceQty = int256(qty) / int256(tkInfo.minimalRecordResolution);
+            totalImbalance += imbalanceQty;
+
+            // add qty overhead
+            extraBps = executeStepFunction(tokenBuyQtySteps[token], int256(qty));
+            rate = addBps(rate, extraBps);
+        } else {
+            // compute token qty
+            imbalanceQty = (-1 * int256(qty)) / int256(tkInfo.minimalRecordResolution);
+            totalImbalance += imbalanceQty;
+
+            // add qty overhead
+            extraBps = executeStepFunction(tokenSellQtySteps[token], int256(qty));
+            rate = addBps(rate, extraBps);
+        }
+
+        if (abs(totalImbalance) >= uint256(tkInfo.maxTotalImbalanceInResolution)) return 0;
+        if (abs(blockImbalance + imbalanceQty) >= uint256(tkInfo.maxPerBlockImbalanceInResolution))
+            return 0;
+
+        return rate;
+    }
+
+    function getBasicRate(IERC20Ext token, bool buy) external view returns (uint256) {
+        if (buy) {
+            return tokenData[token].baseBuyRate;
+        } else {
+            return tokenData[token].baseSellRate;
+        }
+    }
+
+    function getCompactData(IERC20Ext token)
+        external
+        view
+        returns (
+            uint256 arrayIndex,
+            uint256 fieldOffset,
+            int8 buyRateUpdate,
+            int8 sellRateUpdate
+        )
+    {
+        require(tokenData[token].listed, "unlisted token");
+
+        arrayIndex = uint256(tokenData[token].compactDataArrayIndex);
+        fieldOffset = uint256(tokenData[token].compactDataFieldIndex);
+
+        TokenRatesCompactData memory compactData = tokenRatesCompactData[arrayIndex];
+        buyRateUpdate = int8(compactData.buy[fieldOffset]);
+        sellRateUpdate = int8(compactData.sell[fieldOffset]);
+    }
+
+    function getTokenBasicData(IERC20Ext token) external view returns (bool listed, bool enabled) {
+        return (tokenData[token].listed, tokenData[token].enabled);
+    }
+
+    /// @dev for compatible with previous version, this also returns dummy data for imbalance
+    function getStepFunctionData(
+        IERC20Ext token,
+        uint256 command,
+        uint256 param
+    ) external view returns (int256) {
+        if (command == 0 || command == 2) {
+            return int256(tokenBuyQtySteps[token].length);
+        } else if (command == 1) {
+            return tokenBuyQtySteps[token].data[param].x;
+        } else if (command == 3) {
+            return tokenBuyQtySteps[token].data[param].y;
+        } else if (command == 4 || command == 6) {
+            return int256(tokenSellQtySteps[token].length);
+        } else if (command == 5) {
+            return tokenSellQtySteps[token].data[param].x;
+        } else if (command == 7) {
+            return tokenSellQtySteps[token].data[param].y;
+        } else if (command == 8 || command == 10 || command == 12 || command == 14) {
+            return 1;
```

Comment:
seems to me better to  revert here.
with message "imbalance steps not supported."
see how it done in enhanced conversion rates